### DISCUSSION TITLE
Relative computational expense: Material system vs. "hard-coding" + "best practices"

### DISCUSSION THREAD

**Discussion by gka80** [idaholab/moose#17028] [Q&A General] [answered]
Hello-

I had two questions regarding the Materials system that I was hoping someone could answer:

1. What's "best practice" for using the Materials system? I have a tendency to break my materials into smaller "submaterials" or components within my input file (e.g. 'M = M1 + M2 + M3', where 'M1/2/3' represent the subcomponents and 'M' represents the final material that I would pass to a kernel; "usually" these materials are DerivativeParsedMaterial(s)). The benefit to this is that it makes input files easier to debug; however, I'm not sure if this is impacting performance (appreciably). For context, I have ~30 materials, 90% of which are DerivativeParsedMaterials.

2. ...Which leads me to my next question. Would there be any (appreciable) benefit to either: a) Condensing the number of materials used? I suspect that I could reduce the number of materials used by ~50%. Or b) Hard-coding the materials (and their derivatives) into custom material types?

I'm not opposed to either; just trying to understand whether or not I'm abusing the Materials system by using a "ton" of DerivativeParsedMaterials (as opposed to calculating the derivatives by hand and placing them in their own material).

Thank you!

-Garrett

**Comment by cpgr** [comment, marked as answer]
Have you done any [profiling](https://mooseframework.inl.gov/application_development/profiling.html) to see if your materials are taking a significant proportion of the simulation time? If they aren't, it mightn't be worth spending any time refactoring your current design.

**Reply by gka80** [reply]
Thanks for the response.

Good time to learn! I found some small inefficiencies with my code using Instruments: ~20% of the time for the function calls pertain to materials. I'm going to experiment by cleaning up my input file and combining some of the "extra" materials and perform more profiling to see if it is "worthwhile;" hopefully it'll help future users. I'll report back later.

**Reply by mangerij** [reply]
Are you talking about the materials system with automatic differentiation? If so, I think it is definitely slower, possibly by a factor of 2 or 4

**Reply by dschwen** [reply]
@mangerij where does that "factor of 2 or 4" come from? We apply arithmetic optimization and just in time compilation. In my experience these optimizations can create code that evaluates _faster_ than hand coded materials.

**Reply by gka80** [reply]
@dschwen Is this the paper that should be referred to: [https://doi.org/10.1016/j.commatsci.2017.02.017](https://doi.org/10.1016/j.commatsci.2017.02.017)? Are there any instances where the optimizations potentially wouldn't be faster than hand-coded materials?

After doing some more testing, it's unlikely that the performance gains that I would (potentially) experience outweigh the additional commitment to reworking (and testing) the code. I would, however, be interested to hear @mangerij 's experiences.

**Reply by dschwen** [reply]
Yes, that's the paper. It really depends on what your expressions look like. In particular in very long expressions as we see them in thermodynamic free energies there are plenty of opportunities for term reordering, factoring out common factors, replacing float powers with integer powers (big deal) etc. Of course a skilled coder could make these optimization by hand, but the automated system is really good at making them, too. And once the expressions are compiled there is very little additional overhead (nothing form the derivative parsed material except copying variable values of the coupled variables into a fixed size vector to pass it to the JIT code). Any substantial overhead would come from moose calling an additional material class (iterating over the list of material objects, virtual function call overhead). Things that could start to add up if you have _a lot_ of materials that each perform very little work compared to that overhead.

**Reply by mangerij** [reply]
@dschwen just from what I've read in the mailing list... I believe I've seen a few posts regarding this. 

But seeing the timings would make me a believer...

**Reply by mangerij** [reply]
But now I can't find the post. I am pretty sure I saw someone saying AD is much more slow ... Oh well 😂

**Reply by mangerij** [reply]
https://mooseframework.inl.gov/source/systems/NonlinearSystem.html#AD

"One can elect to _sacrifice some computing_ speed and calculate Jacobians automatically using automatic differentiation (AD). MOOSE employs the DualNumber class from the MetaPhysicL package in order to enable AD."

?

**Reply by jessecarterMOOSE** [reply]
Just to be clear, are we talking about the Automatic Differentiation (https://mooseframework.inl.gov/automatic_differentiation/index.html) system for Jacobians, or the unfortunately similarly named Parsed Function-based Automatic Differentiation system (https://mooseframework.inl.gov/modules/phase_field/FunctionMaterials/AutomaticDifferentiation.html). Or are they one in the same? This often confuses me.

**Reply by aeslaughter** [reply]
@jessecarterMOOSE  Two different systems.

**Reply by jessecarterMOOSE** [reply]
Right, good, so I'm not crazy.

So which are we talking about here? @gka80 mentions using DerivativeParsedMaterials, so the latter AD system, while @mangerij mentions Jacobians which is the former AD system? Of course you can use DerivativeParsedMaterials to take derivatives and fill out your Jacobian, meaning you can potentially use both systems.

**Reply by gka80** [reply]
@jessecarterMOOSE I was referring to the latter system. I'm not using AD to form my Jacobians; I've done all of that by hand (aside from using the Cpp interface to pull the derivatives from the material properties). The convergence of the model is quite good, it's just the simulation time (wall time) that's taking awhile. I was looking for any potential options to speed up the simulation; I typically use "quite a few" materials and wanted to know what everyone else's experience was (to make sure I wasn't abusing the system).

After profiling with Instruments, it appears that the heaviest function calls are associated with the preconditioning; the materials (including the composite tensor material); and a custom kernel which I'm working on fixing.
![Screen Shot 2021-02-19 at 16 28 02](https://user-images.githubusercontent.com/30502230/108575685-deff4480-72cf-11eb-9ac9-12445626de17.png)
